### DISCUSSION TITLE
help/minimum_reproducible_example/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#13518] [Docs]
# help/minimum_reproducible_example/

Learn how to create effective Minimum Reproducible Examples (MRE) for bug reports in Ultralytics YOLO repositories. Follow our guide for efficient issue resolution.

https://docs.ultralytics.com/help/minimum_reproducible_example/

**Comment by manoj-samal** [comment]
results= model(im, augment=augment, visualize=visualize)


i'm getting output in tuple format

results = (
tensor[[[[ 2.55792e-01, -8.02494e-01, -1.52477e-02, ..., 1.31591e-01, 7.27416e-01, -7.37097e-01],
[ 4.26252e-01, -5.88908e-01, -2.77992e-01, ..., 5.57948e-02, 4.44748e-01, -7.22842e-01],
[-3.02036e-01, -2.36281e-01, -5.49053e-01, ..., -3.83684e-02, 3.37734e-01, -6.16241e-01]]]]], device='cuda:0'),

(tensor([[[[[ 0.74621, -0.36439, 0.65680, ..., -0.58979, 0.57147, -0.56773],
[-0.17386, -0.39769, 0.95084, ..., -0.61642, 0.48401, -0.48086],
[ 0.01340, -0.13087, 1.09881, ..., -0.53496, 0.38888, -0.53385]
]]]], device='cuda:0'),
[[[[[-5.41013e-02, -2.62636e-02, -4.65444e-02, ..., -5.09703e-02, -4.06154e-02, -7.81466e-02],
[-8.16091e-02, -5.18114e-02, -4.76743e-02, ..., -4.39999e-02, -4.33309e-02, -5.73979e-02],
[-8.58061e-02, -5.21048e-02, -5.23798e-03, ..., -5.07386e-02, -5.69061e-02, -5.13552e-02]]]], device='cuda:0' ))]

   we can seperate it  
   pred , out = results

```
def process_mask(protos, masks_in, bboxes, shape, upsample=False):
    """
    Crop before upsample.
    proto_out: [mask_dim, mask_h, mask_w]
    out_masks: [n, mask_dim], n is number of masks after nms
    bboxes: [n, 4], n is number of masks after nms
    shape:input_image_size, (h, w)

    return: h, w, n
    """

    c, mh, mw = protos.shape  # CHW
    ih, iw = shape
    masks = (masks_in @ protos.float().view(c, -1)).sigmoid().view(-1, mh, mw)  # CHW

    downsampled_bboxes = bboxes.clone()
    downsampled_bboxes[:, 0] *= mw / iw
    downsampled_bboxes[:, 2] *= mw / iw
    downsampled_bboxes[:, 3] *= mh / ih
    downsampled_bboxes[:, 1] *= mh / ih

    masks = crop(masks, downsampled_bboxes)  # CHW
    if upsample:
        masks = F.interpolate(masks[None], shape, mode='bilinear', align_corners=False)[0]  # CHW
    return masks.gt_(0.5)
masks = process_mask(proto[i], det[:, 6:], det[:, :4], im.shape[2:], upsample=True)  # HWC
```

now this masks is in binary

**Reply by pderrenger** [reply]
Hi there! 👋

Thank you for reaching out and providing detailed information about your issue. To help us investigate and resolve this efficiently, could you please provide a Minimum Reproducible Example (MRE)? An MRE is a small, self-contained piece of code that demonstrates the problem you're experiencing. This will help us understand the issue better and work on a fix more effectively.

Here are some steps to create an MRE:

1. **Isolate the Problem**: Remove any unnecessary code or dependencies that are not directly related to the issue.
2. **Use Public Models and Datasets**: Use publicly available models and datasets, such as `yolov8n.pt` and `coco8.yaml`.
3. **Include All Necessary Dependencies**: List the required packages and their versions. You can use `yolo checks` if you have `ultralytics` installed or `pip list` for other tools.
4. **Write a Clear Description of the Issue**: Explain the expected behavior and the actual behavior you're encountering. Include any relevant error messages or logs.
5. **Format Your Code Properly**: Use code blocks to format your code in the issue description.
6. **Test Your MRE**: Ensure that your MRE accurately reproduces the issue and can be run without any modifications.

Here's an example of how you might structure your MRE:

```python
import torch
from ultralytics import YOLO

# Load the model
model = YOLO("yolov8n.pt")

# Load an image
image = torch.rand(1, 3, 640, 640)  # Ensure the image has 3 channels

# Run the model
results = model(image, augment=True, visualize=True)

# Process the results
pred, out = results
print(pred, out)
```

If you haven't already, please ensure you are using the latest versions of `torch` and `ultralytics`. You can upgrade them using:
```bash
pip install --upgrade torch ultralytics
```

For more detailed guidance on creating an MRE, please visit our [Minimum Reproducible Example guide](https://docs.ultralytics.com/help/minimum_reproducible_example).

Looking forward to your response so we can assist you further! 😊

**Comment by manoj-samal** [comment]
i need segmentations in form of [ [670,35][6,305] [60,3]] from segments model of yolov5 where masks are in binary..

**Reply by pderrenger** [reply]
@manoj-samal hello!

Thank you for reaching out. To assist you effectively with obtaining segmentations in the form of `[[670,35],[6,305],[60,3]]` from the YOLOv5 segments model, we'll need a bit more information.

Could you please provide a Minimum Reproducible Example (MRE) that demonstrates the issue you're encountering? This will help us understand the problem better and work towards a solution. You can follow the guidelines here: [Creating a Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example/).

Additionally, please ensure that you are using the latest versions of `torch` and `ultralytics`. You can check your current versions with the following commands:
```bash
pip show torch ultralytics
```
If you are not on the latest versions, please update them:
```bash
pip install --upgrade torch ultralytics
```

For your specific requirement, you can convert binary masks to the desired format using the following approach:

1. **Load the Model and Run Inference:**
    ```python
    from ultralytics import YOLO
    import torch

    # Load the model
    model = YOLO("yolov5-seg.pt")

    # Load an image
    image = torch.rand(1, 3, 640, 640)

    # Run the model
    results = model(image)
    ```

2. **Extract and Convert Masks:**
    ```python
    import numpy as np

    # Assuming results.masks contains the binary masks
    masks = results[0].masks

    # Convert binary masks to coordinates
    def mask_to_coords(mask):
        coords = np.column_stack(np.where(mask > 0))
        return coords.tolist()

    # Example for the first mask
    coords = mask_to_coords(masks[0])
    print(coords)
    ```

Please test this code and let us know if it works for your use case. If you encounter any issues, providing the MRE will be very helpful for us to debug further.

Looking forward to your response! 😊

**Comment by manoj-samal** [comment]
i need segmentations in form of [ [670,35][6,305] [60,3]] from segments model of yolov5 where masks are in binary..
results= model(im, augment=augment, visualize=visualize)

i'm getting output in tuple format

results = (
tensor[[[[ 2.55792e-01, -8.02494e-01, -1.52477e-02, ..., 1.31591e-01, 7.27416e-01, -7.37097e-01],
[ 4.26252e-01, -5.88908e-01, -2.77992e-01, ..., 5.57948e-02, 4.44748e-01, -7.22842e-01],
[-3.02036e-01, -2.36281e-01, -5.49053e-01, ..., -3.83684e-02, 3.37734e-01, -6.16241e-01]]]]], device='cuda:0'),

(tensor([[[[[ 0.74621, -0.36439, 0.65680, ..., -0.58979, 0.57147, -0.56773],
[-0.17386, -0.39769, 0.95084, ..., -0.61642, 0.48401, -0.48086],
[ 0.01340, -0.13087, 1.09881, ..., -0.53496, 0.38888, -0.53385]
]]]], device='cuda:0'),
[[[[[-5.41013e-02, -2.62636e-02, -4.65444e-02, ..., -5.09703e-02, -4.06154e-02, -7.81466e-02],
[-8.16091e-02, -5.18114e-02, -4.76743e-02, ..., -4.39999e-02, -4.33309e-02, -5.73979e-02],
[-8.58061e-02, -5.21048e-02, -5.23798e-03, ..., -5.07386e-02, -5.69061e-02, -5.13552e-02]]]], device='cuda:0' ))]

we can seperate it
pred , out = results

def process_mask(protos, masks_in, bboxes, shape, upsample=False):
    """
    Crop before upsample.
    proto_out: [mask_dim, mask_h, mask_w]
    out_masks: [n, mask_dim], n is number of masks after nms
    bboxes: [n, 4], n is number of masks after nms
    shape:input_image_size, (h, w)

    return: h, w, n
    """

    c, mh, mw = protos.shape  # CHW
    ih, iw = shape
    masks = (masks_in @ protos.float().view(c, -1)).sigmoid().view(-1, mh, mw)  # CHW

    downsampled_bboxes = bboxes.clone()
    downsampled_bboxes[:, 0] *= mw / iw
    downsampled_bboxes[:, 2] *= mw / iw
    downsampled_bboxes[:, 3] *= mh / ih
    downsampled_bboxes[:, 1] *= mh / ih

    masks = crop(masks, downsampled_bboxes)  # CHW
    if upsample:
        masks = F.interpolate(masks[None], shape, mode='bilinear', align_corners=False)[0]  # CHW
    return masks.gt_(0.5)
masks = process_mask(proto[i], det[:, 6:], det[:, :4], im.shape[2:], upsample=True)  # HWC

now this masks is in binary

**Reply by glenn-jocher** [reply]
Hello! 👋

Thank you for reaching out with your query. To assist you effectively, could you please provide a Minimum Reproducible Example (MRE) that demonstrates the issue you're encountering? An MRE is a small, self-contained piece of code that highlights the problem. This helps us understand and resolve the issue more efficiently. You can find a detailed guide on creating an MRE [here](https://docs.ultralytics.com/help/minimum_reproducible_example/).

Additionally, please ensure that you are using the latest versions of `torch` and `ultralytics`. You can check your current versions with the following commands:
```bash
pip show torch ultralytics
```
If they are not up-to-date, please upgrade them and try running your code again:
```bash
pip install --upgrade torch ultralytics
```

Regarding your specific request for segmentation outputs in the form of coordinates, here's a simplified example of how you might process the masks to get the desired format:

```python
import torch
from ultralytics import YOLO

# Load the model
model = YOLO("yolov8n-seg.pt")

# Load an image
image = torch.rand(1, 3, 640, 640)

# Run the model
results = model(image)

# Process the masks
def process_mask(protos, masks_in, bboxes, shape, upsample=False):
    c, mh, mw = protos.shape  # CHW
    ih, iw = shape
    masks = (masks_in @ protos.float().view(c, -1)).sigmoid().view(-1, mh, mw)  # CHW

    downsampled_bboxes = bboxes.clone()
    downsampled_bboxes[:, 0] *= mw / iw
    downsampled_bboxes[:, 2] *= mw / iw
    downsampled_bboxes[:, 3] *= mh / ih
    downsampled_bboxes[:, 1] *= mh / ih

    masks = crop(masks, downsampled_bboxes)  # CHW
    if upsample:
        masks = torch.nn.functional.interpolate(masks[None], shape, mode='bilinear', align_corners=False)[0]  # CHW
    return masks.gt_(0.5)

# Example usage
proto = results[0].protos
det = results[0].boxes
masks = process_mask(proto, det[:, 6:], det[:, :4], image.shape[2:], upsample=True)  # HWC

# Convert binary masks to coordinates (simplified example)
def masks_to_coords(masks):
    coords = []
    for mask in masks:
        y, x = torch.where(mask)
        coords.append(list(zip(x.tolist(), y.tolist())))
    return coords

coords = masks_to_coords(masks)
print(coords)
```

This code snippet demonstrates how to process the masks and convert them into a list of coordinates. Please adjust it according to your specific needs.

Looking forward to your MRE and any additional details you can provide! 😊

**Comment by charming1202** [comment]
from ultralytics import YOLO
import os


def search(model_folder):
    return [os.path.join(model_folder, file) for file in os.listdir(model_folder) if file.endswith(".yaml")]


def train(data_yaml, pre_model, project_dir, model_yamls):
    for model_yaml in model_yamls:
        try:
            print(f"Training with {model_yaml}...")
            model_name = os.path.basename(model_yaml).split('.')[0]
            model = YOLO(model_yaml, task='detect').load(pre_model)
            results = model.train(data=data_yaml,epochs=500,batch=16,imgsz=[1280, 720],workers=4,project=project_dir,name=model_name)
            print(f"Finished training with {model_yaml}")

        except Exception as e:
            print(f"Training for {model_yaml} failed due to {e}. Moving to the next model...")

        finally:
            print(f"Moving to the next model after {model_yaml}.")


if __name__ == '__main__':
    model_folder = r"./yaml_customize/model_yaml"
    model_yamls = search(model_folder)
    data_yaml = r"./yaml_customize/data.yaml"
    pre_model = r"./yolov8n.pt"
    project_dir = r"./tal_topk=5,alpha=0.75,beta=6_ori1280"
    train(data_yaml, pre_model, project_dir, model_yamls)
WARNING ⚠️ updating to 'imgsz=1280' is displayed during training. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640', how to solve it

**Reply by glenn-jocher** [reply]
@charming1202 to resolve the warning about `imgsz=1280`, ensure that the `imgsz` parameter is set as an integer for both training and validation. Adjust your code to use a single integer value like `imgsz=1280` instead of a list. If you need further assistance, please refer to our documentation on creating a Minimum Reproducible Example: https://docs.ultralytics.com/help/minimum_reproducible_example/.

**Reply by charming1202** [reply]
Thank you for your answer, I set imgsz to an int number and I can train normally

**Reply by glenn-jocher** [reply]
You're welcome! I'm glad setting `imgsz` to an integer resolved the issue. If you have any more questions, feel free to ask.
